core/src/Cuda/Kokkos_CudaSpace.cpp

```diff
@@ -172,6 +174,130 @@ void *CudaSpace::allocate(const char *arg_label, const size_t arg_alloc_size,
 }
 
 namespace {
+
+#ifndef CUDART_VERSION
+#error CUDART_VERSION undefined!
+#elif (defined(KOKKOS_ENABLE_IMPL_CUDA_MALLOC_ASYNC) && CUDART_VERSION >= 11020)
+
+/**
+ * If a device has been initialized we will drop it in the set
+ * so we don't initialize it again
+ */
+std::unordered_set<int> _mem_pool_initialized;
+
+/**
+ * This is the initializer. It currently gets the device_id and stream
+ * that impl_alloc gets
+ *    @param device_id  The device id on which to alter the default mempool
+ *    @param stream     A specific stream for cudaMallocAsync
+ * 		@param error_code return any failure codes for cuda APIs
+ *    @return  				  true for success, false for failure
+ */
+bool initializeMempool(const int device_id, const cudaStream_t stream,
+                       cudaError_t &error_code) {
+  error_code = cudaSuccess;
+
+  // Interpret env var
+  char *env_string = getenv("EXPERIMENTAL_CUDA_ASYNC_MEMPOOL_SIZE");
+  if (!env_string) return true;  // Nothing to set, initialization is complete
+
+  // Env var can be a decimal e.g. 2.4g
+  double factor = 1;  // bytes
+
+  // Interpret the string
+  std::string mempool_size_string(env_string);
+  char last_letter = mempool_size_string.back();
+
+  // We can deal with the last letter being units.
+  if (std::isalpha(last_letter)) {
+    mempool_size_string.pop_back();
+    char last_letter_cap = std::toupper(last_letter);
+    switch (last_letter_cap) {
+      case 'K': factor = 1024; break;
+      case 'M': factor = 1024 * 1024; break;
+      case 'G': factor = 1024 * 1024 * 1024; break;
+      default:
+        std::cerr << "Unknown units for requested size " << last_letter << "\n";
+        return false;
+        break;
+    }
```

Comment:
Also, this function feels like it's the "opposite" of https://github.com/kokkos/kokkos/blob/662043966274936f2c7a367e58975f06c2814cf7/core/src/impl/Kokkos_Error.cpp#L47-L71, so it might be nice to be consistent with the notations ? (I mean `B`, `KiB` and so on)